# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/interfaces/base/nsIBrowser.idl
# Commit: d027f301f08c
# Full Hash: d027f301f08c38d8ab3095abfe0a04f2255f6e6c
# Author: Nika Layzell <nika@thelayzells.com>
# Date: 2020-06-16 03:30:48
# Regressor Bug: 1640019
# File Overlap Count: 1
# Description:
#   Bug 1640019 - Part 1: Support toplevel process switches outside of tabbrowser, r=mattwoodrow
#   
#   This new process switching behavior is only enabled for some browser elements,
#   which have specified a specific attribute. Turning this on for all browsers with
#   a `remote` attribute causes breakage in reftests.
# ==============================================================================

diff -r 82becb1f6eae -r d027f301f08c dom/interfaces/base/nsIBrowser.idl
--- a/dom/interfaces/base/nsIBrowser.idl	Tue Jun 16 02:46:15 2020 +0300
+++ b/dom/interfaces/base/nsIBrowser.idl	Mon Jun 15 23:23:43 2020 +0000
@@ -161,12 +161,42 @@
                                in AString aContentType);
 
   /**
+   * Determine what process switching behavior this browser element should have.
+   */
+  cenum ProcessBehavior : 8 {
+    // Gecko won't automatically change which process this frame, or it's
+    // subframes, are loaded in.
+    PROCESS_BEHAVIOR_DISABLED,
+
+    // If `useRemoteTabs` is enabled, Gecko will change which process this frame
+    // is loaded in automatically, without calling `performProcessSwitch`.
+    // When `useRemoteSubframes` is enabled, subframes will change processes.
+    PROCESS_BEHAVIOR_STANDARD,
+
+    // When `useRemoteTabs` is enabled, Gecko will call `performProcessSwitch`
+    // in order to change which process this frame is loaded in.
+    // When `useRemoteSubframes` is enabled, subframes will change processes.
+    PROCESS_BEHAVIOR_CUSTOM,
+
+    // Gecko won't automatically change which process this frame is loaded, but
+    // when `useRemoteSubframes` is enabled, subframes will change processes.
+    //
+    // NOTE: This configuration is included only for backwards compatibility,
+    // and will be removed, as it can easily lead to invalid behavior.
+    PROCESS_BEHAVIOR_SUBFRAME_ONLY,
+  };
+  readonly attribute nsIBrowser_ProcessBehavior processSwitchBehavior;
+
+  /**
    * Called by Gecko when it wants to change the process which is currently
    * being used to render content in this tab.
    *
    * Returns a promise which will be resolved with the ChildID of the new
    * process.
    *
+   * NOTE: This method is only called if `processSwitchBehavior` is
+   * `PROCESS_BEHAVIOR_CUSTOM`.
+   *
    * @param aRemoteType the new remote type to switch this browser into
    * @param aPendingSwitchId unique identifier for the ongoing
    *                         process-switching load
@@ -177,7 +207,4 @@
   Promise performProcessSwitch(in AString aRemoteType,
                                in uint64_t aPendingSwitchId,
                                in boolean aReplaceBrowsingContext);
-
-  /** If `true`, this browser supports the `performProcessSwitch` method */
-  readonly attribute bool canPerformProcessSwitch;
 };